CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,348 @@
+			   (pf.particleId() == reco::PFCandidate::ParticleType::egamma_HF && longE < longFibreThr_))
+			  toKill = true;
+
+			LogDebug("PFCandidateRecalibrator") << "====> ieta,iphi,depth: " 
+							    << badIt.ieta << " " << badIt.iphi << " " << badIt.depth << " corr: " << badIt.ratio;
+			LogDebug("PFCandidateRecalibrator") << "====> recal en (tot,H,E): " 
+							    << totEnergy << " " << hcalEnergy << " " << ecalEnergy;
+
+		      }
+		  }
+		
+		if(toKill)
+		  {
+		    discarded->push_back(pf);
+		    oldToNew[i] = (-discarded->size());
+		    badToOld.push_back(i);
+
+		    LogDebug("PFCandidateRecalibrator") << "==> KILLED ";
+		  }
+		else
+		  {
+		    copy->push_back(pf);
+		    oldToNew[i] = (copy->size());
+		    newToOld.push_back(i);
+		    
+		    copy->back().setHcalEnergy(hcalEnergy, hcalEnergy);
+		    copy->back().setEcalEnergy(ecalEnergy, ecalEnergy);
+
+		    float scalingFactor = totEnergy/totEnergyOrig;
+		    math::XYZTLorentzVector recalibP4 = pf.p4() * scalingFactor;
+		    copy->back().setP4( recalibP4 );
+
+		    LogDebug("PFCandidateRecalibrator") << "====> stored en (tot,H,E): " 
+							<< copy->back().energy() << " " << copy->back().hcalEnergy() << " " << copy->back().ecalEnergy();
+		  }
+	      }
+	    else
+	      {
+		copy->push_back(pf);
+		oldToNew[i] = (copy->size());
+		newToOld.push_back(i);
+	      }
+	  }
+	else
+	  {
+	    copy->push_back(pf);
+	    oldToNew[i] = (copy->size());
+	    newToOld.push_back(i);
+	  }
+      }
+
+    // Now we put things in the event
+    edm::OrphanHandle<reco::PFCandidateCollection> newpf = iEvent.put(std::move(copy));
+    edm::OrphanHandle<reco::PFCandidateCollection> badpf = iEvent.put(std::move(discarded), "discarded");
+
+    std::unique_ptr<edm::ValueMap<reco::PFCandidateRef>> pf2pf(new edm::ValueMap<reco::PFCandidateRef>());
+    edm::ValueMap<reco::PFCandidateRef>::Filler filler(*pf2pf);
+    std::vector<reco::PFCandidateRef> refs; refs.reserve(nPfCand);
+
+    // old to new
+    for (auto iOldToNew : oldToNew){
+      if (iOldToNew > 0) {
+	refs.push_back(reco::PFCandidateRef(newpf, iOldToNew-1));
+      } else {
+	refs.push_back(reco::PFCandidateRef(badpf,-iOldToNew-1));
+      }
+    }
+    filler.insert(pfcandidates, refs.begin(), refs.end());
+    // new good to old
+    refs.clear();
+    for (int i : newToOld) {
+      refs.push_back(reco::PFCandidateRef(pfcandidates,i));
+    }
+    filler.insert(newpf, refs.begin(), refs.end());
+    // new bad to old
+    refs.clear();
+    for (int i : badToOld) {
+      refs.push_back(reco::PFCandidateRef(pfcandidates,i));
+    }
+    filler.insert(badpf, refs.begin(), refs.end());
+    // done
+    filler.fill();
+    iEvent.put(std::move(pf2pf));
+}
+
+void
+PFCandidateRecalibrator::fillDescriptions(edm::ConfigurationDescriptions& descriptions)
+{
+  edm::ParameterSetDescription desc;
+
+  desc.add<edm::InputTag>("pfcandidates");
+  desc.add<double>("shortFibreThr", 1.4);
+  desc.add<double>("longFibreThr", 1.4);
+
```

Comment:
this is missing to pass desc out.

```descriptions.add("pfCandidateRecalibrator", desc);```
this will lead to the system making a default pfCandidateRecalibrator_cfi.py file, which can be used directly.

The default instance in this PR is called ```pfCandidatesBadHadRecalibrated```. It seems reasonable.
However, then PFCandidateRecalibrator is non-specific and will clash with possible other future recalibrators.
So, it makes sense to rename this producer to PFCandidateBadHadRecalibrator and the instance to be pfCandidateBadHadRecalibrator

Reply:
We could also leave the PFCandidateRecalibrator in place, assuming that future versions will just add more functionality.
In that case, the name of the module instance in the workflow better stay similarly generic pfCandidatesBadHadRecalibrated